File: DSA/Data Structure/6_trees/bst.py
from __future__ import annotations

# 2. Binary Search Tree (BST) ==============================================================================
# What: A binary tree where left.val < node.val < right.val.
# Why: Enables O(log n) search/insert/delete (if balanced).

# Must‑know operations:
# insert, search, delete
# Validate BST
# Kth smallest/largest element
# Floor/ceil, range queries

# Practice:
# LeetCode 98: Validate BST
# LeetCode 230: Kth Smallest BST
# LeetCode 450: Delete Node in a BST

class BST:
    def __init__(self, data: int):
        self.data = data
        self.left: BST = None
        self.right: BST = None
        
    def insert(self, data: int): 
        if not data:
            return
        
        if self.data == data:
            return
        
        if data < self.data:
            if self.left:
                self.left.insert(data)
            else:
                self.left = BST(data)
            
        else:
            if self.right:
                self.right.insert(data)
            else:
                self.right = BST(data)
                
    def inorder(self:BST):
        def _inorder(node: BST):
            if node:
                _inorder(node.left)
                print(node.data)
                _inorder(node.right)
        return _inorder(self)
    
    def preorder(self: BST):
        def _preorder(node: BST):
            if not node:
                return
            print(node.data)
            _preorder(node.left)
            _preorder(node.right)
        return _preorder(self)
    
    def search(self: BST, data:int):
        def _search(node: BST, data: int):
            if not node:
                return False
            if data == node.data:
                return True
            elif data < node.data:
                return  _search(node.left, data)
            else:
                return _search(node.right, data)
        return _search(self, data)
    
    def delete(self: BST, data: int):
        pass
                    
bst = BST(21)
bst.insert(2)
bst.insert(4)
bst.insert(3)
bst.insert(5)
bst.insert(29)
bst.insert(11)
bst.insert(23)
bst.insert(5)
print("INSERT ENDS!")

bst.inorder()

# print(bst.search(23))

File: DSA/DS_TYPESCRIPT/7_TOP_K_ELEMENTS/bst.ts
`
Given the root of a binary search tree and an integer k, return the kth largest value in the BST.
Step-by-Step Collection of Top 3 Largest Elements

           11
         /    \
       6       16
     /  \    /    \
    3    9  14    19
   / \  / \ / \   /  \
  2  5 8 10 13 15 18 20
 /      /      /     /
1      12     17    -

`

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val: number) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

function topKLargestInBST(root: TreeNode | null, k: number): number[] {
    const result: number[] = [];
    function reverseInorder(node: TreeNode | null) {
        if (!node || result.length >= k) return;
        reverseInorder(node.right);
        if (result.length < k) {
            result.push(node.val);
            reverseInorder(node.left);
        }
    }
    reverseInorder(root);
    return result;
}
